### PR TITLE
fix(platform/gitlab): handle assignee id not found

### LINKED ISSUE #27943: "failed to add assignee (Cannot read properties of undefined (reading 'id'))" on Gitlab self-hosted
### Discussed in https://github.com/renovatebot/renovate/discussions/27890

<div type='discussions-op-text'>

<sup>Originally posted by **deronnax** March 12, 2024</sup>
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Gitlab

### Please tell us more about your question or problem

Hello.
I run renovate on one of our repo, and on this repository, renovate has an assignee problem: it fails to add assignees to an MR (see logs).
I won't be able to do a minimum reproducible use case, but maybe we can work something else.
I am pretty sure it did work when using reviewers instead of assignees.
Thank you very much.

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: Adding assignees 'orga/core/squad-honeytoken' to #11290 (repository=gg-code/prm/ward-runs-app, branch=renovate/python-gitlab-4.x)
DEBUG: addAssignees error (repository=gg-code/prm/ward-runs-app, branch=renovate/python-gitlab-4.x)
       "err": {
         "message": "Cannot read properties of undefined (reading 'id')",
         "stack": "TypeError: Cannot read properties of undefined (reading 'id')\n    at getUserID (/usr/local/renovate/lib/modules/platform/gitlab/http.ts:10:13)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at Proxy.addAssignees (/usr/local/renovate/lib/modules/platform/gitlab/index.ts:1196:24)\n    at addParticipants (/usr/local/renovate/lib/workers/repository/update/pr/participants.ts:61:11)\n    at ensurePr (/usr/local/renovate/lib/workers/repository/update/pr/index.ts:501:9)\n    at processBranch (/usr/local/renovate/lib/workers/repository/update/branch/index.ts:753:28)\n    at writeUpdates (/usr/local/renovate/lib/workers/repository/process/write.ts:166:17)\n    at update (/usr/local/renovate/lib/workers/repository/process/extract-update.ts:214:11)\n    at Object.renovateRepository (/usr/local/renovate/lib/workers/repository/index.ts:77:19)\n    at attributes.repository (/usr/local/renovate/lib/workers/global/index.ts:200:11)\n    at start (/usr/local/renovate/lib/workers/global/index.ts:185:7)\n    at /usr/local/renovate/lib/renovate.ts:18:22"
       }
 WARN: Failed to add assignees (repository=gg-code/prm/ward-runs-app, branch=renovate/python-gitlab-4.x)
```

</details>
</div>

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- throw a better error message incase ID for an assignee is not found
- do not fail when one assignee not found, add the others

## Context

Closes: https://github.com/renovatebot/renovate/issues/27943
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/gitlab/index.ts [hunk 1]
```diff
@@ -1217,7 +1217,12 @@ export async function addAssignees(
     logger.debug(`Adding assignees '${assignees.join(', ')}' to #${iid}`);
     const assigneeIds: number[] = [];
     for (const assignee of assignees) {
-      assigneeIds.push(await getUserID(assignee));
+      try {
+        const userId = await getUserID(assignee);
+        assigneeIds.push(userId);
+      } catch (err) {
+        logger.warn({ assignee }, 'Failed to add assignee');
```
Comment (rarkins) [inline lib/modules/platform/gitlab/index.ts:1224]:
```suggestion
        logger.debug({ assignee, err }, 'getUserID() error');
        logger.warn({ assignee }, 'Failed to add assignee - could not get ID');
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/platform/gitlab/http.ts [hunk 2]
```diff
@@ -1,13 +1,22 @@
...
 export const gitlabApi = new GitlabHttp();
 
 export async function getUserID(username: string): Promise<number> {
-  return (
+  const userInfo = (
     await gitlabApi.getJson<{ id: number }[]>(`users?username=${username}`)
-  ).body[0].id;
+  ).body;
+
+  if (!is.nonEmptyArray(userInfo)) {
```
Comment (viceice) [inline lib/modules/platform/gitlab/http.ts:13]: I would invert the condition for cleaner code

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/platform/gitlab/http.ts (hunk 2 above)
Comment (secustor) [reply lib/modules/platform/gitlab/http.ts:13]:
```suggestion
  if (is.emptyArray(userInfo)) {
```

**Review by rarkins — APPROVED** [review]

**Pull request merged**